Simplifying implementation of `ExpressibleByStringLiteral`

therealbnut opened the pull request:
<!-- What's in this pull request? -->

**Added default implementations for `ExpressibleByUnicodeScalarLiteral` and `ExpressibleByExtendedGraphemeClusterLiteral`**
```swift
init(unicodeScalarLiteral value: UnicodeScalar)
init(extendedGraphemeClusterLiteral value: Character)
```

Now implementing `ExpressibleByStringLiteral` is as simple as:
```swift
extension Example: ExpressibleByStringLiteral {
    init(stringLiteral value: String) {
        self.string = value
    }
}
```

Previously you would have to implement something like this:
```swift
extension Example: ExpressibleByStringLiteral {
    init(stringLiteral value: String) {
        self.string = value
    }
    init(extendedGraphemeClusterLiteral value: Character) {
        self.string = String(describing: value)
    }
    init(unicodeScalarLiteral value: UnicodeScalar) {
        self.string = String(describing: Character(value))
    }
}
```
As well as understand what each of these meant, and which types can be used.

**Changed `Character`'s implementation of `ExpressibleByUnicodeScalarLiteral`**
```diff
-  public init(unicodeScalarLiteral value: Character) {
+  public init(unicodeScalarLiteral value: UnicodeScalar) {
```
The documentation states:
```swift
/// Valid types for `UnicodeScalarLiteralType` are `UnicodeScalar`,
/// `String`, and `StaticString`.
```

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
<!-- Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN). -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

therealbnut commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

therealbnut commented:
thanks :)

therealbnut commented:
This fails on source stability, this was expected, I've pushed a commit to remove the specific issue from the PR, although perhaps the documentation/test should be updated instead (See PR description).

```
swift/test/api-digester/tmp/source-stability.swift.expected
+Constructor Character.init(unicodeScalarLiteral:) has 1st parameter type change from Character to UnicodeScalar
```

xwu commented on stdlib/public/core/CompilerProtocols.swift:
```diff
@@ -395,6 +395,14 @@ public protocol ExpressibleByExtendedGraphemeClusterLiteral
 }
 
+extension ExpressibleByExtendedGraphemeClusterLiteral
+    where Self.ExtendedGraphemeClusterLiteralType == Character
```
Drive-by nit: indents are two spaces in this file; otherwise a delightful addition!

therealbnut commented:
Fixed by e8e5240 thanks @xwu

therealbnut commented:
Also fixed indentation and **`where`** with d187c34, I left it indented after a search in stdlib for `^\s*where`.

CodaFi commented:
Is there documentation mentioning implementation requirements that requires updating here?

therealbnut commented:
@CodaFi not that I can find, the documentation states:

> To add `ExpressibleByStringLiteral` conformance to your custom type, implement **the** required initializer.

It already implies you only need to implement `init(stringLiteral:)`.

There is a comment that could possibly be removed.
```swift
  // FIXME: when we have default function implementations in protocols, provide
  // an implementation of init(extendedGraphemeClusterLiteral:).
```

therealbnut committed:
Simplify conforming to ExpressibleByStringLiteral with default implementations

therealbnut commented:
ping @dabrahams (as suggested by moiseev)

therealbnut commented:
ping @dabrahams (`"If it is not urgent, the common courtesy ping rate is one week."` let me know if I'm pestering you, or I should be pestering someone else)

dabrahams commented:
@airspeedswift let's discuss this.  It's clearly an improvement, barely squeaks by on not needing evolution review IMO, but we are planning to redesign everything in this space.  I guess I vote for merging if it passes tests.

dabrahams commented:
@swift-ci Please smoke test

dabrahams commented on stdlib/public/core/CompilerProtocols.swift:
```diff
@@ -338,7 +338,7 @@ public protocol ExpressibleByUnicodeScalarLiteral {
   ///
   /// Valid types for `UnicodeScalarLiteralType` are `UnicodeScalar`,
-  /// `String`, and `StaticString`.
+  /// `Character`, `String`, and `StaticString`.
```
Please add a test that verifies the correctness of this edit.

therealbnut replied:
Should be addressed by https://github.com/apple/swift/pull/7125/files#diff-4d7f8659b54e15ede9391722ae11cb74R24

Was this what you had in mind?

I‘m just verifying the tests locally, then I'll ping you so we can run them with CI and finalise the PR.

dabrahams replied:
It's a start, but it doesn't check semantics.  Please write a test using StdlibUnittest that checks the results at runtime, thanks!

therealbnut replied:
I believe this is now covered by https://github.com/apple/swift/pull/7125/files/2201e882fa4715afb2b941bc86ed3ef991cc5eae..e67a2b3439b27d2d31d4b6a9745cbf1664002ed0 thanks for the feedback :)

therealbnut replied:
@dabrahams hi, just pinging you to hopefully finalise this review :)

therealbnut replied:
@dabrahams hey, pinging again to try and get this merged :)

dabrahams reviewed (changes requested):
Please add tests that verify your default implementations work.  Also please eliminate any concrete implementations that are redundant with these defaults, except where it hurts performance.

Thanks!

airspeedswift commented:
@dabrahams I agree, can go ahead w/o evolution, and may as well merge even if we end up redesigning just in case that misses 4.0 due to other higher-priority stuff

therealbnut committed:
attributes on default implementations

therealbnut committed:
ExpressibleByUnicodeScalarLiteral validation test

therealbnut committed:
more generic default implementations

therealbnut committed:
clean up test

therealbnut committed:
remove unneeded implementations

therealbnut committed:
remove test verification

therealbnut commented:
**note: I haven't had a chance to run the tests locally yet**

I believe I've addressed all the comments. I'm not 100% sure if the tests are in the right place though.

I had to make the protocol defaults slightly more generalised to remove redundant concrete implementations, the existing concrete implementations often used the same type for each protocol. I believe the generalisation improved the code overall though.

Implementing a valid `associatedtype` on one of these protocols will always produce a valid `associatedtype` in the default implementations. This is because the allowed types on each protocol is a subset of the types in its parent protocol. 

| protocol `associatedtype` | one of |
| --- | --- |
| `UnicodeScalarLiteralType` | `String`, `StaticString`, `Character`, `UnicodeScalar` |
| `ExtendedGraphemeClusterLiteralType` | `String`, `StaticString`, `Character` |
| `StringLiteralType` | `String`, `StaticString` |

therealbnut commented:
I'm getting an error trying to build master, I'll run tests locally once I've worked out what's going on there.
```bash
swift/include/swift/AST/TypeAlignments.h:104:21: error: no template named 'AlignOf' in namespace 'llvm'; did you mean 'AlignTo'?
static_assert(llvm::AlignOf<void*>::Alignment >= 2,
              ~~~~~~^~~~~~~
                    AlignTo
```

therealbnut committed:
indent

jrose-apple commented on stdlib/public/core/CompilerProtocols.swift:
```diff
@@ -396,9 +396,12 @@ public protocol ExpressibleByExtendedGraphemeClusterLiteral
-    self.init(extendedGraphemeClusterLiteral: Character(value))
+where ExtendedGraphemeClusterLiteralType == UnicodeScalarLiteralType {
+
+  @effects(readonly)
```
Please don't specify this. You can't guarantee the initializer you chain to follows it.

therealbnut replied:
good point, thank you

therealbnut committed:
revert @effects and affected methods

therealbnut committed:
fix test generics with _ protocols

therealbnut committed:
Add semantic tests

therealbnut committed:
clean up tests

therealbnut committed:
Fix redundant conformance requirements

airspeedswift commented:
@swift-ci Please test

therealbnut commented:
@airspeedswift Thanks for running CI.

The **failing test** seems to be in [swift-package-manager](https://github.com/apple/swift-package-manager/blob/master/Tests/BasicTests/ProcessSetTests.swift), I don't see how this could be affected by this branch.

I've merged latest master into this branch (it was 20 days behind with no conflicts), it would be awesome if you could run CI again, unless you think it may be this branch at fault.

airspeedswift commented:
@swift-ci Please test

therealbnut commented:
@airspeedswift sorry to pester you again, CI failed again, it claims "wrong sha", I suspect it's because it ran on 4d3ca09 and not 78c5df1...

I rebased on master, perhaps that will fix it...

airspeedswift commented:
@swift-ci Please test

airspeedswift commented:
c'mon @swift-ci you can do it...

therealbnut commented:
@airspeedswift I'm not sure if that actually triggered a build, the currently running build was triggered half an hour ago, and this one was 7 hours ago...

airspeedswift commented:
@swift-ci Please test

therealbnut commented:
Hmm, it seems to have used the wrong commit again

airspeedswift commented:
Ok it's down to an unrelated failure on the full macOS test suite that I think is also failing in master. We can smoke test then merge probably.

airspeedswift commented:
@swift-ci Please smoke test macOS platform

therealbnut commented:
Thank @airpseedswift :)

therealbnut commented:
Thanks @airspeedswift 🙂

therealbnut commented:
Thank you everyone that helped review :)

therealbnut commented:
Thank you @airspeedswift for your persistence with CI :)
